File: python/hash-table/word-pattern.py
class Solution(object):
    def wordPattern(self, pattern, str):
        """
        :type pattern: str
        :type str: str
        :rtype: bool
        """
        ls = str.split()
        d1 = {}
        d2 = {}
        if(len(ls) != len(pattern)):
            return False
        for c1, c2 in zip(pattern, ls):
            if c1 not in d1:
                d1[c1] = c2
            if c2 not in d2:
                d2[c2] = c1
            if d1[c1] != c2 or d2[c2] != c1:
                return False

        return True

File: c/hash-table/word-pattern.c
int is_pattern(char *d[], char c, char *word)
{
    if(strcmp(d[c-'a'], word) != 0)
        return false;

    for(int i = 0; i < 26; ++i)
        if(d[i] && strcmp(d[i], word) == 0 && 'a' + i != c)
            return false;

    return true;
}

bool wordPattern(char* pattern, char* str) {
    char *d[26];
    char *sep = " ";
    char *word;
    int i = 0;
    char c;
    int len = strlen(pattern);
    memset(d, 0, sizeof(char *) * 26);
    word = strtok(str, sep);
    while(word != NULL){
        if(i >= len)
            return false;
        c = pattern[i++];
        if(d[c - 'a'] == NULL)
            d[c - 'a'] = strdup(word);

        if(!is_pattern(d, c, word))
            return false;
        word = strtok(NULL, sep);
    }

    if(i != len)
        return false;

    return true;
}
